docs/devel/kubectl-conventions.md

```diff
@@ -135,6 +133,151 @@ Updated: 8/27/2015
 * Use "TYPE" for the particular flavor of resource type accepted by kubectl, rather than "RESOURCE" or "KIND"
 * Use "NAME" for resource names
 
+## Command implementation conventions
+
+For every command there should be a `NewCmd<CommandName>` function that creates the command and returns a pointer to a `cobra.Command`, which can later be added to other parent commands to compose the structure tree. There should also be a `<CommandName>Options` struct with a variable to every flag and argument declared by the command (and any other variable required for the command to run). This makes tests and mocking easier. The struct ideally exposes three functions:
+
+* `Complete`: Completes the struct variables with values that may not be directly provided, for example, by flags pointers. Here you will usually take the `args` slice and set the values as appropriate variables, instantiate configs or clients, etc.
+* `Validate`: performs validation and returns errors.
+* `Run<CommandName>`: runs the actual command, taking as assumption that the struct is complete with all required values to run, and they are valid.
```

Comment:
The separation between Complete, Validate, and Run makes cleaner where the business logic will exist or most probably called from (Run), and where cobra, factory, etc. fall into (Complete). There are a bunch of commands in OpenShift (https://github.com/openshift/origin/issues/3697) that already use this pattern and a couple of [kubectl](https://github.com/kubernetes/kubernetes/pull/17366#issuecomment-159894781) commands too